### PR TITLE
[Dispatch] Fold `collapse(expand)` unit dims

### LINKED ISSUE #19263: Llama 3.1 8B fp16 TP8 sharded fails to compile for CPU and GPU
### What happened?

When I try to compile the sharded Llama 3.1 8b fp16 [IR](https://gist.github.com/aviator19941/bab5886f53f2fd0b3b8458519148542c) for CPU or GPU:

I get this error for CPU:
https://gist.github.com/aviator19941/82bceb2624571d446da0964440790fde

and this error for GPU:
https://gist.github.com/aviator19941/89761b3bbb6ace5a6945de667e6d1e39

I tried to use these flags that were suggested to be used when compiling Llama as well:
`
--iree-dispatch-creation-enable-aggressive-fusion=true --iree-global-opt-propagate-transposes=true --iree-opt-aggressively-propagate-transposes=true --iree-opt-data-tiling=false --iree-preprocessing-pass-pipeline='builtin.module(util.func(iree-preprocessing-generalize-linalg-matmul-experimental))' --iree-hal-indirect-command-buffers=true --iree-stream-resource-memory-model=discrete --iree-hip-legacy-sync=false --iree-hal-memoization=true --iree-opt-strip-assertions
`

### Steps to reproduce your issue

1. wget the IR: https://gist.github.com/aviator19941/bab5886f53f2fd0b3b8458519148542c
2. Try to compile for CPU:
`
../iree-build-no-trace/tools/iree-compile 8b_f16_tp8_decomposed.mlir -o=8b_f16_tp8_decomposed_cpu.vmfb --iree-hal-target-device=llvm-cpu[0] --iree-hal-target-device=llvm-cpu[1] --iree-hal-target-device=llvm-cpu[2] --iree-hal-target-device=llvm-cpu[3] --iree-hal-target-device=llvm-cpu[4] --iree-hal-target-device=llvm-cpu[5] --iree-hal-target-device=llvm-cpu[6] --iree-hal-target-device=llvm-cpu[7]
`
3. CPU error: https://gist.github.com/aviator19941/82bceb2624571d446da0964440790fde
4. Try to compile for GPU:
`
../iree-build-no-trace/tools/iree-compile 8b_f16_tp8_decomposed.mlir --iree-hip-target=gfx942 -o=8b_f16_tp8_decomposed.vmfb --iree-hal-target-device=hip[0] --iree-hal-target-device=hip[1] --iree-hal-target-device=hip[2] --iree-hal-target-device=hip[3] --iree-hal-target-device=hip[4] --iree-hal-target-device=hip[5] --iree-hal-target-device=hip[6] --iree-hal-target-device=hip[7]
`
5. GPU error: https://gist.github.com/aviator19941/89761b3bbb6ace5a6945de667e6d1e39

### What component(s) does this issue relate to?

_No response_

### Version information

iree-base-compiler 3.1.0rc20241121

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by IanWood1**
Fold unit dimensions for the pattern `collapse(expand)` by removing any unnecessary unit dims that are expanded out and subsequently collapsed again. This prevents the collapse/expand ops, which contain unneeded unit dims, from being propagated and introducing unit dims back to linalg ops.

-----

Doesn't quite resolve https://github.com/iree-org/iree/issues/19263 because there are `flow.tensor.bitcast` ops that are preventing the unit dims from being completely folded:

```mlir
%expanded_24 = tensor.expand_shape %214 [[0, 1, 2], [3]] output_shape [4, %21, 1, 128] : tensor<?x128xf16> into tensor<4x?x1x128xf16>
%323 = flow.tensor.bitcast %expanded_24 : tensor<4x?x1x128xf16>{%21} -> tensor<4x?x1x64xcomplex<f16>>{%21}
%collapsed_40 = tensor.collapse_shape %323 [[0], [1, 2], [3]] : tensor<4x?x1x64xcomplex<f16>> into tensor<4x?x64xcomplex<f16>>
```

### CODE REVIEW

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 1]
```diff
@@ -22,6 +22,8 @@
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
 #include "mlir/Dialect/Linalg/Transforms/Transforms.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/IR/Verifier.h"
```
Comment (IanWood1) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: remove after CI run

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 2]
```diff
@@ -32,6 +33,146 @@ namespace mlir::iree_compiler::DispatchCreation {
...
+      for (auto [innerIdx, inDim] : llvm::enumerate(indicies)) {
+        // Can't drop this dim if it isnt statically 1 or if it isn't being
+        // combined with any other dimensions.
+        if (indicies.size() == 1 || interShape[inDim] != 1) {
+          continue;
+        }
+
+        // If we are collapsing multiple unit dims together, at least 1 must be
+        // kept (prefer the first).
+        if (outShape[outDim] == 1 && innerIdx != 0) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp:72]: I think I understand what you are trying to do, but I dont follow the logic that much. Maybe some comments would help?

**Review by MaheshRavishankar — COMMENTED** [review]
I trust this works fine (the tests etc. looks good), and I overall follow what you are checking for, but maybe more comments would help here... these things are hard to follow.

**Comment by benvanik** [comment]
Please rebase if this is not stale.

**Comment by IanWood1** [comment]
This is stale and I haven't ran into this issue again. Will revisit if it pops back up
